I118 integrate viva

clevengr opened the pull request:
### Description of what the PR does
   Integrates the VIVA Input Validator into PC2, also fixing some issues with the existing Custom Input Validator code so that both VIVA and Custom Input Validators operate in a uniform manner.

### Issue which the PR fixes
Fixes #118.  The PR includes 15 new JUnit tests which test both the "VivaAdapter" class (the class used to adapt VIVA to PC2) as well as testing new ContestSnakeYAMLLoader support for loading Input Validator configurations (both VIVA and Custom Input Validator configurations) via YAML.  The PR also includes an updated version of the PC2 Admin Guide with Appendix O (Input Validators) updated to describe both how to use VIVA, how to use Custom Input Validators, and how to specify Input Validator configurations via YAML files.

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
  Windows 8.1 and 10 with Java version "1.8.0_201"

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)
See the attached file [PR.i118.TestSteps.txt](https://github.com/pc2ccs/pc2v9/files/5432486/PR.i118.TestSteps.txt) for complete testing steps for both interactive Input Validator configuration as well as testing YAML loading of Input Validator configurations.

troy2914 commented:
Ugh, so many file changes,  presumably NL endings...

clevengr commented:
Yes, I'd like to avoid those NL changes... any suggestions as to how?

john


On Sat, Oct 24, 2020 at 9:57 AM Troy <notifications@github.com> wrote:

> Ugh, so many file changes, presumably NL endings...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pc2ccs/pc2v9/pull/124#issuecomment-715993835>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGT544O2ULFJQCMK5HVP63SMMBNLANCNFSM4S5LY7CQ>
> .
>

clevengr commented:
@troy2914 Would the "NL endings" problem(s) you referred to (above) be fixed by the outstanding PR #111 ?

troy2914 commented:
I am not sure, though that PR is out of date now.

clevengr commented:
Why is PR #111 "out of date now"?

troy2914 commented:
> Why is PR #111 "out of date now"?

it's showing conflicts of a few files.  And it would be re-done after this was committed.

clevengr commented:
By "after *this *was committed"... does "*this*" refer to PR #124?



On Mon, Oct 26, 2020 at 6:00 PM Troy <notifications@github.com> wrote:

> Why is PR #111 <https://github.com/pc2ccs/pc2v9/pull/111> "out of date
> now"?
>
> it's showing conflicts of a few files. And it would be re-done after this
> was committed.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pc2ccs/pc2v9/pull/124#issuecomment-716909422>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGT545Z7XPRHSF3K3MTUCTSMYLTBANCNFSM4S5LY7CQ>
> .
>

lane55 commented:
Per Troy's comment about NL, there are tens of thousands of lines to review if the NL are not fixed.
Other than fixing the NL directly, is there a way for github to ignore NL changes.  I doubt it.

clevengr commented:
> Per Troy's comment about NL, there are tens of thousands of lines to review if the NL are not fixed.
> Other than fixing the NL directly, is there a way for github to ignore NL changes. I doubt it.

My impression was that the way to do this (force github to "ignore NL changes") is to tell GIT to always store files with Unix newlines, and then to configure it so that when you check out a file it converts the NLs to the local platform (and correspondingly, when you check a file back IN it converts the NLs to Unix form).

It's also my impression that this is what PR #111 was intended to do?   This is one of the things I'd like to go over in a meeting...

troy2914 commented:
Looking better.   Please continue with the rest of the NL changing files.

troy2914 commented:
src diffs now look sane,  please fix the test ones.

clevengr commented:
I believe I have finished fixing all the issues with NLs in the PR files.  Note the following:

- Two files under the "test" folder (**ui/EditProblemPaneTest** and **core/model/ProblemTest**) still show "100% of file lines changed -- _even though I converted them using dos2unix and they show as "unix" files on my machine and I have pushed them_.  I cannot explain this, unless it's some kind of delay on GitHub...
- There are still other files in my local machine Git repo that show as "Windows" files.  However, I believe that none of them affect reviewing the PR.
- The PR shows "153 files modified".  However, I should point out that _over 100 of those files_ are new "test data" files which were added under /testdata to support the 15 new JUnit tests I added for testing VIVA input validator operations.  You can spend your time reviewing all those test data files if you want, but I'm not sure why....

Please report if you see any other issues, or if you have any idea why **ProblemTest** and **EditProblemPaneTest** are still showing as "100% replacements" even though they are now in Unix format...

troy2914 commented:
ProblemTest.java and EditProblemPaneTest.java were in DOS (with CRLF line terminators) previously, not unix.

clevengr commented:
Hmmm.... so I presume this is an issue we'll have to live with, short-term,
and which will go away once we get PR #111 completely implemented?


On Sat, Oct 31, 2020 at 5:48 PM Troy <notifications@github.com> wrote:

> ProblemTest.java and EditProblemPaneTest.java were in DOS (with CRLF line
> terminators) previously, not unix.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pc2ccs/pc2v9/pull/124#issuecomment-720008882>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGT544VLD7GXBHUAOL3IX3SNSV33ANCNFSM4S5LY7CQ>
> .
>

troy2914 commented:
for right now you should change those 2 files to DOS file format,  and all these NL stuff will disappear with #111

troy2914 commented:
The diff looks sane now.   Seems primary area to test, is to ensure the UI is properly detecting changes, and that 
input and output validators still work.

clevengr commented:
The "test list" which I attached to the PR includes steps for testing
various UI changes; I think I covered all cases but of course I welcome
people trying other tests to see if they can make it misbehave.  I also
tried all combinations of VIVA input validation (with both valid and
invalid patterns) plus two different versions of a Custom Input Validator
(both of which are checked in, in the "testdata" folder).  I did NOT try
any special OUTPUT validation tests (other than all the normal ones which
run with JUnits); I'm not aware that I made any changes that would affect
Ouput Validation (but of course it's good to check/test).

And BTW, the answer to a question which came up, which several of us
*thought* we knew the answer but no one was certain:  "yes, when you push a
change to a fork which has a PR, GitHub automatically updates the files in
the PR".  I didn't have to do anything, after running dos2unix on the
offending files, other than "push" to my fork; GitHub took care of updating
the PR.  Very nice.


On Sat, Oct 31, 2020 at 10:55 PM Troy <notifications@github.com> wrote:

> The diff looks sane now. Seems primary area to test, is to ensure the UI
> is properly detecting changes, and that
> input and output validators still work.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pc2ccs/pc2v9/pull/124#issuecomment-720037175>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGT546IQBFS5LFKRDMJPJTSNTZ35ANCNFSM4S5LY7CQ>
> .
>

lane55 commented on src/edu/csus/ecs/pc2/ui/DefineCustomInputValidatorPane.java:
```diff
@@ -0,0 +1,350 @@
+            } else {
+                System.err.println ("Internal error: grandparent of DefineCustomInputValidatorPane is not an EditProblemPane; cannot enable Add/Update button"
+                        + "\nPlease report this problem to the PC2 Development Team (pc2@ecs.csus.edu");
+                getController().getLog().severe("grandparent of DefineCustomInputValidatorPane is not an EditProblemPane");
```
Should log message and displayed messages be identical?

clevengr replied:
I don't see why this should necessarily be the case; displayed messages are generally for the user sitting at the console, while log messages are, generally speaking, more for the systems person who's going to pore through logs looking for detailed information which in fact might not make any sense to the average user at the console.

I will concede however that the systems person looking at the log might also like to see the "Internal error" part of the message, along with the "cannot enable Add/Update button"; I'll push that as a change to the PR.

lane55 commented on test/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoaderTest.java:
```diff
@@ -1286,7 +1287,254 @@ public void testValidatorKeys() throws Exception {
 
     }
 
-    // SOMEDAY get this JUnit working
```
Is this a SOMEDAY or a TODO ?

clevengr replied:
First, aren't ALL "TODOs" effectively "someday"?   If it's something "to be done", it's by definition NOT currently done, so it "needs to be done", and that could be later today, tomorrow, next week, next month, next year... at some point in the future -- i.e. "someday".
Second, we do not have, have never had, any formal specification of what anything like "TODO" or "SOMEDAY" means in a comment.  Not saying we wouldn't benefit by having such; just saying we do not have any such thing and never have had any such thing.  So asking "Is this a SOMEDAY or a TODO" is pretty much a moot question since there's nothing anywhere that defines what either of those formally means.
Third, **I did not put this comment in the code**; it was there when I started working on this PR. 

I don't think this is an important issue, at least not with regard to this PR.  If you disagree, feel free to file a new Issue on it.

lane55 reviewed (changes requested):
1 - In Problem class, adding fields (String [] and SF) to (flyweight) Problem may bloat/change class size.
Note that all SF and other large data structures are currently stored in ProblemDataFiles.

2 - per comment, is the SOMEDAY (aka likely never) be changed to TODO ?

clevengr commented on src/edu/csus/ecs/pc2/core/model/Problem.java:
```diff
@@ -141,18 +183,33 @@
+    //VIVA Input validator settings - only relevant if the problem was saved with Viva Input Validator Settings
+    private boolean problemHasVivaInputValidatorPattern = false;
+    private boolean vivaInputValidatorHasBeenRun = false;
+    private String [] vivaInputValidatorPattern = null;
```
Most of the VIVA patterns I have seen are very small -- on the order of 5-10 lines of code.  I think it's reasonable to keep the pattern as a String in the Problem class.

clevengr replied:
If you disagree, I suggest filing this as a new Issue.

clevengr commented on src/edu/csus/ecs/pc2/core/model/Problem.java:
```diff
@@ -390,6 +461,7 @@ public String toString() {
     public String toStringDetails() {
         String retStr = "Problem[";
         
+        //basic configuration settings
```
First comment:  I don't think Apache Commons is currently part of our vendor library; making this change would require adding that library, which I think it rather large and seems like unwarranted bloat for such a small gain.
Second comment: I'll assume that since you said "Someday", you don't consider this an item that's necessary to fix now to get the PR approved.
On that basis, I'm marking the comment "Resolved".

clevengr commented on testdata/InputValidatorTest/SumitInputValidator.java:
```diff
@@ -0,0 +1,78 @@
+import java.io.BufferedReader;
```
I suppose it's possible this could happen, although the class is pretty simple so it seems unlikely.  I don't think this is worth holding up the approval of the PR; if you think it's necessary I think it should be filed as a separate Issue.

clevengr commented on src/edu/csus/ecs/pc2/ui/RunInputValidatorsPane.java:
```diff
@@ -235,23 +236,54 @@ protected void populateGUI() {
-
-                    result.addResult(res); // do we need to clone the InputValidationResult?
+            //check if the problem has a Custom Input Validator
+            if (probs[prob].getCurrentInputValidatorType()==INPUT_VALIDATOR_TYPE.CUSTOM) {
```
To each their own style; your approach would require executing an extra runtime statement to assign the value of probs[prob] to the variable "problem" so that the variable could then be used 'more clearly' (in your opinion).  I'm not convinced that's a worthwhile tradeoff -- but again, to each their own style.

I'm assuming you don't think this rises to the level of rejecting the PR because of it.  If that's not the case, feel free "unresolve" the conversation and/or add a new Issue for it.  (But I don't see anyone getting around to spending their valuable time on fixing such a trivial item...)

clevengr commented:
> 1 - In Problem class, adding fields (String [] and SF) to (flyweight) Problem may bloat/change class size.
> Note that all SF and other large data structures are currently stored in ProblemDataFiles.
> 
> 2 - per comment, is the SOMEDAY (aka likely never) be changed to TODO ?

I will make the first requested change(s) and push an update.
Regarding the second requested change, see my comments on that particular comment.

Finally, I think there is an error in process here.  First, I got a message "laned Approved these changes".  If I hadn't been bogged down with other things and had acted on that message immediately, I would have taken "Approval" as meaning it was OK to merge the change (since I'm pretty sure that's what the intent of "Approving" is supposed to be).  As it is, I didn't get to processing the "Approval" before I got a subsequent message "Changes Requested".  I believe that "requesting changes" is incompatible with "Approval".  If you want changes, request them; if you're willing to accept the PR as is, Approve it -- but don't do both.

At least, that's how I think the process is supposed to work...

lane55 commented:
re: Approval to Changes requested.    I think that worked properly, I approved, then after a walk I re-viewed
the PR and the Problem and saw the flyweight class was potentially no longer flyweight.   

We should consider a refactor for the existing String [] (testCaseDataFilenames, testCaseAnswerFilenames) fields 
moving the contents out of the Problem class, in part because it is not scalable and potentially bloats the Problem.

lane55 commented:
Should it hold up the PR, no, I'll grant this exception.  Furture code
  should use foreach for reasons below.   

  Yes foreach this is a sytle question, and yes in my professional opinion it
  is more clear, consistent with most other code in the system, in thie case
  the overhead is negligible esp in this case, and less error prone.   

On Fri, Nov 20, 2020 at 11:23:18AM -0800, John Clevenger wrote:
> @clevengr commented on this pull request.
> 
> 
> 
> >      private Vector<ProblemInputValidationResults> getInputValidationResultsTableData(Problem [] probs) {
>          
>          Vector<ProblemInputValidationResults> tableData = new Vector<ProblemInputValidationResults> ();
>          
>          for (int prob=0; prob<probs.length; prob++) {
> -            if (probs[prob].isProblemHasInputValidator()) {
> -                
> -                ProblemInputValidationResults result = new ProblemInputValidationResults(probs[prob]);
> -                
> -                // add each result for the current problem to the current row
> -                for (InputValidationResult res : probs[prob].getInputValidationResults()) {
> -
> -                    result.addResult(res); // do we need to clone the InputValidationResult?
> +            //check if the problem has a Custom Input Validator
> +            if (probs[prob].getCurrentInputValidatorType()==INPUT_VALIDATOR_TYPE.CUSTOM) {
> 
> To each their own style; your approach would require executing an extra runtime statement to assign the value of probs[prob] to the variable "problem" so that the variable could then be used 'more clearly' (in your opinion).  I'm not convinced that's a worthwhile tradeoff -- but again, to each their own style.
> 
> I'm assuming you don't think this rises to the level of rejecting the PR because of it.  If that's not the case, feel free "unresolve" the conversation and/or add a new Issue for it.  (But I don't see anyone getting around to spending their valuable time on fixing such a trivial item...)
> 
> 
> -- 
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://github.com/pc2ccs/pc2v9/pull/124#discussion_r527922914

lane55 commented:
Not a show stopper, just a note.  


On Fri, Nov 20, 2020 at 11:29:12AM -0800, John Clevenger wrote:
> @clevengr commented on this pull request.
> 
> 
> 
> > +     * the user to save the problem changes.
> +     */
> +    private void enableUpdateButton() {
> +        JPanePlugin parent = getParentPane();
> +        if (parent != null && parent instanceof InputValidatorPane) {
> +            final JPanePlugin grandParent = ((InputValidatorPane)parent).getParentPane();
> +            if (grandParent != null && grandParent instanceof EditProblemPane) {
> +                SwingUtilities.invokeLater( new Runnable() {
> +                    public void run () {
> +                        ((EditProblemPane)grandParent).enableUpdateButton();
> +                    }
> +               });
> +            } else {
> +                System.err.println ("Internal error: grandparent of DefineCustomInputValidatorPane is not an EditProblemPane; cannot enable Add/Update button"
> +                        + "\nPlease report this problem to the PC2 Development Team (pc2@ecs.csus.edu");
> +                getController().getLog().severe("grandparent of DefineCustomInputValidatorPane is not an EditProblemPane");
> 
> I don't see why this should necessarily be the case; displayed messages are generally for the user sitting at the console, while log messages are, generally speaking, more for the systems person who's going to pore through logs looking for detailed information which in fact might not make any sense to the average user at the console.
> 
> I will concede however that the systems person looking at the log might also like to see the "Internal error" part of the message, along with the "cannot enable Add/Update button"; I'll push that as a change to the PR.
> 
> -- 
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://github.com/pc2ccs/pc2v9/pull/124#discussion_r527925676

lane55 commented:
Yes all TODOs, XXX, etc. are sometday - my point is that there is 
  no agreement about how to use these notes and when they should
  be done or how urgent they are.

  The question is _when_ should someday/todo occur?   

  At work either all TODOs, etc. are fixed and removed or there
  is some way to determine what the priority of this todo is.

  JUnits should not be broken or commented out ad nauseam.  They should
  be written, removed or scheduled to be written (aka put issue number
  into the comment showing when they are fixed, for example).

On Fri, Nov 20, 2020 at 11:35:09AM -0800, John Clevenger wrote:
> @clevengr commented on this pull request.
> 
> 
> 
> > @@ -1286,7 +1287,254 @@ public void testValidatorKeys() throws Exception {
>  
>      }
>  
> -    // SOMEDAY get this JUnit working
> 
> First, aren't ALL "TODOs" effectively "someday"?   If it's something "to be done", it's by definition NOT currently done, so it "needs to be done", and that could be later today, tomorrow, next week, next month, next year... at some point in the future -- i.e. "someday".
> Second, we do not have, have never had, any formal specification of what anything like "TODO" or "SOMEDAY" means in a comment.  Not saying we wouldn't benefit by having such; just saying we do not have any such thing and never have had any such thing.  So asking "Is this a SOMEDAY or a TODO" is pretty much a moot question since there's nothing anywhere that defines what either of those formally means.
> Third, **I did not put this comment in the code**; it was there when I started working on this PR. 
> 
> I don't think this is an important issue, at least not with regard to this PR.  If you disagree, feel free to file a new Issue on it.
> 
> 
> -- 
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://github.com/pc2ccs/pc2v9/pull/124#discussion_r527928766

lane55 commented:
What do you disagree with ?

  I cannot find what this refers to, so at this time I need
  more information before filing an issue or requesting
  the change in the PR.


On Fri, Nov 20, 2020 at 12:06:15PM -0800, John Clevenger wrote:
> @clevengr commented on this pull request.
> 
> 
> 
> >      
> -    private Vector<InputValidationResult> inputValidationResults ;
> +    //custom input validator settings - only relevant if the problem was saved with a Custom Input Validator
> +    private boolean problemHasCustomInputValidator = false;
> +    private boolean customInputValidatorHasBeenRun = false;
> +//    private String customInputValidatorProgramName = "";  //program name should be determined by the Custom Input Validator Serialized File
> +    private String customInputValidatorCommandLine = "";
> +//    private String customInputValidatorFilesOnDiskFolderName = "";   
> +    private SerializedFile customInputValidatorSerializedFile = null;
> +    private InputValidationStatus customInputValidationStatus = InputValidationStatus.UNKNOWN;
> +    private Vector<InputValidationResult> customInputValidationResults = null;
> +
> +    //VIVA Input validator settings - only relevant if the problem was saved with Viva Input Validator Settings
> +    private boolean problemHasVivaInputValidatorPattern = false;
> +    private boolean vivaInputValidatorHasBeenRun = false;
> +    private String [] vivaInputValidatorPattern = null;
> 
> If you disagree, I suggest filing this as a new Issue.
> 
> -- 
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://github.com/pc2ccs/pc2v9/pull/124#discussion_r527943279

lane55 commented:
Agreed.  If the pattern is changed to a String, I approve.

  Next time, I recommend encapsulating the Viva fields in a
  VivaProblemSettings class because a Problem flyweight is a flyweight,
  So only having a single class reference is better than 15+ more fields,
  also using the class makes it very clear that the fields only refer
  to Viva rather than having some affect on a Problem.  

  In the future I will not approve a PR if fields are added that
  can be naturally put into a class/encapsulated.  This isn't COBOL :)

On Fri, Nov 20, 2020 at 11:08:13AM -0800, John Clevenger wrote:
> @clevengr commented on this pull request.
> 
> 
> 
> >      
> -    private Vector<InputValidationResult> inputValidationResults ;
> +    //custom input validator settings - only relevant if the problem was saved with a Custom Input Validator
> +    private boolean problemHasCustomInputValidator = false;
> +    private boolean customInputValidatorHasBeenRun = false;
> +//    private String customInputValidatorProgramName = "";  //program name should be determined by the Custom Input Validator Serialized File
> +    private String customInputValidatorCommandLine = "";
> +//    private String customInputValidatorFilesOnDiskFolderName = "";   
> +    private SerializedFile customInputValidatorSerializedFile = null;
> +    private InputValidationStatus customInputValidationStatus = InputValidationStatus.UNKNOWN;
> +    private Vector<InputValidationResult> customInputValidationResults = null;
> +
> +    //VIVA Input validator settings - only relevant if the problem was saved with Viva Input Validator Settings
> +    private boolean problemHasVivaInputValidatorPattern = false;
> +    private boolean vivaInputValidatorHasBeenRun = false;
> +    private String [] vivaInputValidatorPattern = null;
> 
> Most of the VIVA patterns I have seen are very small -- on the order of 5-10 lines of code.  I think it's reasonable to keep the pattern as a String in the Problem class.
> 
> -- 
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://github.com/pc2ccs/pc2v9/pull/124#discussion_r527914986

clevengr commented:
I have pushed an update to PR #124 which moves all VIVA variables and results out of the `Problem `class and into a new class `VivaInputValidatorSettings`.  The `Problem` class now contains just a single variable related to VIVA -- a reference to a `ViviInputValidatorSettings` object; all accessors which previously referenced VIVA data stored in `Problem` now delegate to methods in the `VivaInputValidatorSettings` object.  

The updated code passes all JUnits (including the ones which were added as part of the PR explicitly to test VIVA, plus all the pre-existing JUnits).   I reran the complete set of Interactive and Batch Tests described in the file  `PR.i118.TestSteps.txt` attached to the original PR #124 Description; all of them pass as expected.

Requesting re-review of the PR and (hopefully) approval.

lane55 reviewed (approved):
Approved.

lane55 commented on src/edu/csus/ecs/pc2/core/model/Problem.java:
```diff
@@ -141,18 +183,33 @@
+//    private String customInputValidatorProgramName = "";  //program name should be determined by the Custom Input Validator Serialized File
+    private String customInputValidatorCommandLine = "";
+//    private String customInputValidatorFilesOnDiskFolderName = "";   
+    private SerializedFile customInputValidatorSerializedFile = null;
```
Postpone this change to be included in a future refactor.

lane55 commented on src/edu/csus/ecs/pc2/core/model/Problem.java:
```diff
@@ -141,18 +183,33 @@
+    private boolean vivaInputValidatorHasBeenRun = false;
+    private String [] vivaInputValidatorPattern = null;
+    private InputValidationStatus vivaInputValidationStatus = InputValidationStatus.UNKNOWN;
+    private Vector<InputValidationResult> vivaInputValidationResults = null;
```
Instead of ProblemDataFile, put VIVA settings in VivaInputValidatorSettings.   VIVA settings will be stored
in Problem, until a refactor moves their storage outside of Problem similar to how ProblemDataFiles are stored.

lane55 replied:
With move of all Viva settings into new VivaInputValidatorSettings class, this item is resolved
